src/include/processor/operator/recursive_extend/bfs_state.h

```diff
@@ -0,0 +1,153 @@
+#pragma once
+
+#include "processor/operator/mask.h"
+
+namespace kuzu {
+namespace processor {
+
+enum VisitedState : uint8_t {
+    NOT_VISITED_DST = 0,
+    VISITED_DST = 1,
+    NOT_VISITED = 2,
+    VISITED = 3,
+};
+
+struct Frontier {
+    std::vector<common::offset_t> nodeOffsets;
+    // Multiplicity stands for number of paths that can reach an offset.
+    std::unordered_map<common::offset_t, uint64_t> offset2Multiplicity;
```

Comment:
2 things: 
1) I'd call this nodeMultiplicity. Also, I'd rather not adopt the style of using 2 (for "to") and 4 (for "for).
2) This is only needed by VariableLengthJoin and not by ShortestPathRecursiveJoin. We should not inject it into ShortestPathRecursiveJoin. Maybe you can template BaseBFSMorsel with the type of Frontier it has and have a Frontier and FrontierWithMultiplicity. Something along those lines?